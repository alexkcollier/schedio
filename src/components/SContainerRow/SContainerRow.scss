.row {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin: -1 * spacing();

  & + & {
    margin-top: spacing();
  }

  &--justify {
    &-start {
      @include generate-all-breakpoints {
        justify-content: flex-start;
      }
    }

    &-center {
      @include generate-all-breakpoints {
        justify-content: center;
      }
    }

    &-end {
      @include generate-all-breakpoints {
        justify-content: flex-end;
      }
    }

    &-around {
      @include generate-all-breakpoints {
        justify-content: space-around;
      }
    }

    &-between {
      @include generate-all-breakpoints {
        justify-content: space-between;
      }
    }
  }

  &--align {
    &-start {
      @include generate-all-breakpoints {
        align-items: flex-start;
      }
    }

    &-center {
      @include generate-all-breakpoints {
        align-items: center;
      }
    }

    &-end {
      @include generate-all-breakpoints {
        align-items: flex-end;
      }
    }

    &-around {
      @include generate-all-breakpoints {
        align-items: space-around;
      }
    }

    &-between {
      @include generate-all-breakpoints {
        align-items: space-between;
      }
    }
  }

  &--reverse {
    flex-direction: row-reverse;
  }
}
