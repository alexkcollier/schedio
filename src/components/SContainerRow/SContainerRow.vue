<template>
  <div
    :class="classList"
    class="row"
  >
    <slot />
  </div>
</template>

<script>
import ContainerMethods from '@/mixins/ContainerMethods'
import rowOptions from './row-options'
import mobileBreakpoints from '@/components/SContainer/mobile-breakpoints'

const flexOptions = Object.values(rowOptions)
const validator = v => (flexOptions.includes(v) ? true : console.error(`Allowed: ${flexOptions}`))

const propSettings = () => ({
  type: String,
  default: '',
  validator
})

// set up all $props relying on mobile breakpoints
const breakpointProps = mobileBreakpoints.reduce((props, breakpoint) => {
  props[`align-after-${breakpoint}`] = propSettings()
  props[`align-until-${breakpoint}`] = propSettings()
  props[`justify-after-${breakpoint}`] = propSettings()
  props[`justify-until-${breakpoint}`] = propSettings()

  return props
}, {})

export default {
  name: 'SContainerRow',

  mixins: [ContainerMethods],

  props: {
    ...breakpointProps,

    isReversed: {
      type: Boolean,
      default: false
    },

    /* eslint-disable vue/require-default-prop */
    // eslint doesn't see prop defaults in functions
    // Add base props without breakpoints
    justify: propSettings(),

    align: propSettings()
    /* eslint-enable */
  },

  computed: {
    theNames() {
      return this.generateResponsiveClassNames('align|justify')
    },

    classList() {
      return [
        this.align ? `row--align-${this.align}` : '',
        this.justify ? `row--justify-${this.justify}` : '',
        this.isReversed ? `row--reverse` : '',
        ...this.generateResponsiveClassNames('align|justify')
      ]
    }
  },

  methods: {
    // used by mixin
    classNameReducer: classes => (acc, key) => {
      const rx = /(align|justify)-(.+)/
      // e.g.: $props.alignUntilXl => 'row--align-center-until-xl'
      const name = key.replace(rx, `row--$1-${classes[key]}-$2`)

      if (classes[key] && key.match(rx)) return acc.concat(name)

      return acc
    }
  }
}
</script>
