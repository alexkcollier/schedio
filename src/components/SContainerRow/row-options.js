export default {
  Unset: '',
  start: 'start',
  center: 'center',
  end: 'end',
  around: 'around',
  between: 'between',
};
