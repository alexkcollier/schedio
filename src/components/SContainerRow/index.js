import SContainerRow from './SContainerRow.vue'

export { SContainerRow }

export default { install: Vue => Vue.component(SContainerRow.name, SContainerRow) }
