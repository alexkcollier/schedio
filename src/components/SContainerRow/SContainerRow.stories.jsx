import { SContainer } from '@/components/SContainer'
import { SContainerColumn } from '@/components/SContainerColumn'
import { SContainerRow } from '@/components/SContainerRow'
import rowOptions from '@/components/SContainerRow/row-options'
import { SHeading } from '@/components/SHeading'
import { boolean, select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import orderBy from 'lodash.orderby'
import PropList from '@/docs/PropList'
import kebabCase from 'lodash.kebabcase'

storiesOf('Components/Layout.SContainer', module)
  .addParameters({ jest: 'SContainerRow' })
  .add('Container Row', function SContainerRowDocs() {
    // get sorted row props from component
    const rowPropNames = orderBy(Object.keys(SContainerRow.props))

    // flexbox align items
    const alignProps = rowPropNames.reduce((p, name) => {
      return /^(align)/.test(name)
        ? { ...p, [name]: select(kebabCase(name), rowOptions, '', 'Alignment') }
        : p
    }, {})

    // flexbox justify content
    const justifyProps = rowPropNames.reduce((props, name) => {
      return /^(justify)/.test(name)
        ? { ...props, [name]: select(kebabCase(name), rowOptions, '', 'Justification') }
        : props
    }, {})

    return {
      props: {
        props: {
          default: () => ({
            ...alignProps,
            ...justifyProps,
            isReversed: boolean('is-reversed', false, 'Other')
          })
        }
      },
      render(h) {
        const { props } = this.$props

        return (
          <div>
            <SHeading level="1">Rows</SHeading>
            <p>
              The container hides the padding from its parent container, and provides a flexbox
              container for <code>SContainerColumn</code>. It has been styled for visibility, and is
              normally invisible.
            </p>
            <p>
              Vertical alignment of row children can be set with the <code>align</code> props, and
              horizontal justification can be set with the <code>justify</code> props. They can also
              be reversed by adding the <code>is-reversed</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>

            <SContainer style="border: 1px solid black;">
              The container
              <SContainerRow {...{ props }} style="background-color: silver;">
                <SContainerColumn>Some content</SContainerColumn>
                <SContainerColumn>Some more content</SContainerColumn>
              </SContainerRow>
            </SContainer>

            <PropList component={SContainerRow} />
          </div>
        )
      }
    }
  })
