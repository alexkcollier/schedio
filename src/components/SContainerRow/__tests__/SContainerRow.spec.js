import mobileBreakpoints from '@/components/SContainer/mobile-breakpoints'
import SContainerRow from '@/components/SContainerRow/SContainerRow.vue'
import rowOptions from '@/components/SContainerRow/row-options'
import { shallowMount } from '@vue/test-utils'
import camelCase from 'lodash.camelcase'

describe('SContainerRow.vue', () => {
  const staticProps = ['justify', 'align']
  // Generate the same props generated in `SContainerRow.vue`
  const propsToGenerate = [
    'align-until',
    'align-after',
    'justify-until',
    'justify-after',
  ]
  const propList = propsToGenerate.reduce((list, prop) => {
    mobileBreakpoints.forEach(breakPoint => list.push(camelCase(`${prop}-${breakPoint}`)))
    return list
  }, staticProps)

  const rowPropTest = (propName, value) => {
    wrapper.setProps({ [propName]: value })
    expect(wrapper.props(propName)).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  }

  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SContainerRow, {
      slots: {
        default: '<div>Row Content</div>',
      },
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  // eslint-disable-next-line jest/expect-expect
  it('can be reversed', () => {
    rowPropTest('isReversed', true)
  })

  it('generates props for mobile breakpoints', () => {
    propList.forEach(expect(wrapper.props()).toHaveProperty)
  })

  propList.forEach((propName) => {
    const flexOptions = Object.values(rowOptions).filter(opt => !!opt)
    flexOptions.forEach((option) => {
      // eslint-disable-next-line jest/expect-expect
      it(`can be ${propName} ${option}`, () => rowPropTest(propName, option))
    })
  })
})
