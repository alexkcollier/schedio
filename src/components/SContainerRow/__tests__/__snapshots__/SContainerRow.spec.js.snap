// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SContainerRow.vue can be align around 1`] = `
<div class="row row--align-around">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be align between 1`] = `
<div class="row row--align-between">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be align center 1`] = `
<div class="row row--align-center">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be align end 1`] = `
<div class="row row--align-end">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be align start 1`] = `
<div class="row row--align-start">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterLg around 1`] = `
<div class="row row--align-around-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterLg between 1`] = `
<div class="row row--align-between-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterLg center 1`] = `
<div class="row row--align-center-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterLg end 1`] = `
<div class="row row--align-end-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterLg start 1`] = `
<div class="row row--align-start-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterMd around 1`] = `
<div class="row row--align-around-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterMd between 1`] = `
<div class="row row--align-between-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterMd center 1`] = `
<div class="row row--align-center-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterMd end 1`] = `
<div class="row row--align-end-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterMd start 1`] = `
<div class="row row--align-start-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterSm around 1`] = `
<div class="row row--align-around-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterSm between 1`] = `
<div class="row row--align-between-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterSm center 1`] = `
<div class="row row--align-center-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterSm end 1`] = `
<div class="row row--align-end-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterSm start 1`] = `
<div class="row row--align-start-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterXl around 1`] = `
<div class="row row--align-around-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterXl between 1`] = `
<div class="row row--align-between-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterXl center 1`] = `
<div class="row row--align-center-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterXl end 1`] = `
<div class="row row--align-end-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignAfterXl start 1`] = `
<div class="row row--align-start-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilLg around 1`] = `
<div class="row row--align-around-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilLg between 1`] = `
<div class="row row--align-between-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilLg center 1`] = `
<div class="row row--align-center-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilLg end 1`] = `
<div class="row row--align-end-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilLg start 1`] = `
<div class="row row--align-start-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilMd around 1`] = `
<div class="row row--align-around-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilMd between 1`] = `
<div class="row row--align-between-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilMd center 1`] = `
<div class="row row--align-center-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilMd end 1`] = `
<div class="row row--align-end-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilMd start 1`] = `
<div class="row row--align-start-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilSm around 1`] = `
<div class="row row--align-around-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilSm between 1`] = `
<div class="row row--align-between-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilSm center 1`] = `
<div class="row row--align-center-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilSm end 1`] = `
<div class="row row--align-end-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilSm start 1`] = `
<div class="row row--align-start-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilXl around 1`] = `
<div class="row row--align-around-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilXl between 1`] = `
<div class="row row--align-between-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilXl center 1`] = `
<div class="row row--align-center-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilXl end 1`] = `
<div class="row row--align-end-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be alignUntilXl start 1`] = `
<div class="row row--align-start-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justify around 1`] = `
<div class="row row--justify-around">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justify between 1`] = `
<div class="row row--justify-between">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justify center 1`] = `
<div class="row row--justify-center">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justify end 1`] = `
<div class="row row--justify-end">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justify start 1`] = `
<div class="row row--justify-start">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterLg around 1`] = `
<div class="row row--justify-around-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterLg between 1`] = `
<div class="row row--justify-between-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterLg center 1`] = `
<div class="row row--justify-center-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterLg end 1`] = `
<div class="row row--justify-end-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterLg start 1`] = `
<div class="row row--justify-start-after-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterMd around 1`] = `
<div class="row row--justify-around-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterMd between 1`] = `
<div class="row row--justify-between-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterMd center 1`] = `
<div class="row row--justify-center-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterMd end 1`] = `
<div class="row row--justify-end-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterMd start 1`] = `
<div class="row row--justify-start-after-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterSm around 1`] = `
<div class="row row--justify-around-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterSm between 1`] = `
<div class="row row--justify-between-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterSm center 1`] = `
<div class="row row--justify-center-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterSm end 1`] = `
<div class="row row--justify-end-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterSm start 1`] = `
<div class="row row--justify-start-after-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterXl around 1`] = `
<div class="row row--justify-around-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterXl between 1`] = `
<div class="row row--justify-between-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterXl center 1`] = `
<div class="row row--justify-center-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterXl end 1`] = `
<div class="row row--justify-end-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyAfterXl start 1`] = `
<div class="row row--justify-start-after-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilLg around 1`] = `
<div class="row row--justify-around-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilLg between 1`] = `
<div class="row row--justify-between-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilLg center 1`] = `
<div class="row row--justify-center-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilLg end 1`] = `
<div class="row row--justify-end-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilLg start 1`] = `
<div class="row row--justify-start-until-lg">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilMd around 1`] = `
<div class="row row--justify-around-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilMd between 1`] = `
<div class="row row--justify-between-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilMd center 1`] = `
<div class="row row--justify-center-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilMd end 1`] = `
<div class="row row--justify-end-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilMd start 1`] = `
<div class="row row--justify-start-until-md">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilSm around 1`] = `
<div class="row row--justify-around-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilSm between 1`] = `
<div class="row row--justify-between-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilSm center 1`] = `
<div class="row row--justify-center-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilSm end 1`] = `
<div class="row row--justify-end-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilSm start 1`] = `
<div class="row row--justify-start-until-sm">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilXl around 1`] = `
<div class="row row--justify-around-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilXl between 1`] = `
<div class="row row--justify-between-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilXl center 1`] = `
<div class="row row--justify-center-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilXl end 1`] = `
<div class="row row--justify-end-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be justifyUntilXl start 1`] = `
<div class="row row--justify-start-until-xl">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue can be reversed 1`] = `
<div class="row row--reverse">
  <div>Row Content</div>
</div>
`;

exports[`SContainerRow.vue renders correctly 1`] = `
<div class="row">
  <div>Row Content</div>
</div>
`;
